drivers/SmartThings/matter-lock/src/new-matter-lock/init.lua

```diff
@@ -0,0 +1,1670 @@
+  end
+end
+
+local function delete_user_from_table(device, userIdx)
+  -- If User Index is ALL_INDEX, remove all entry from the table
+  if userIdx == ALL_INDEX then
+    device:emit_event(capabilities.lockUsers.users({}, {visibility = {displayed = false}}))
+    return
+  end
+
+  -- Get latest user table
+  local user_table = utils.deep_copy(device:get_latest_state(
+    "main",
+    capabilities.lockUsers.ID,
+    capabilities.lockUsers.users.NAME,
+    {}
+  ))
+
+  -- Re-create user table
+  for index, entry in pairs(user_table) do
+    if entry.userIndex == userIdx then
+      table.remove(user_table, index)
+      break
+    end
+  end
+  device:emit_event(capabilities.lockUsers.users(user_table, {visibility = {displayed = false}}))
+end
+
+----------------------
+-- Credential Table --
+----------------------
+local function add_credential_to_table(device, userIdx, credIdx, credType)
+  -- Get latest credential table
+  local cred_table = utils.deep_copy(device:get_latest_state(
+    "main",
+    capabilities.lockCredentials.ID,
+    capabilities.lockCredentials.credentials.NAME,
+    {}
+  ))
+
+  -- Add new entry to table
+  table.insert(cred_table, {userIndex = userIdx, credentialIndex = credIdx, credentialType = credType})
+  device:emit_event(capabilities.lockCredentials.credentials(cred_table, {visibility = {displayed = false}}))
+end
+
+local function delete_credential_from_table(device, credIdx)
+  -- If Credential Index is ALL_INDEX, remove all entries from the table
+  if credIdx == ALL_INDEX then
+    device:emit_event(capabilities.lockCredentials.credentials({}))
+  end
+
+  -- Get latest credential table
+  local cred_table = utils.deep_copy(device:get_latest_state(
+    "main",
+    capabilities.lockCredentials.ID,
+    capabilities.lockCredentials.credentials.NAME,
+    {}
+  ))
+
+  -- Delete an entry from credential table
+  local userIdx = 0
+  for index, entry in pairs(cred_table) do
+    if entry.credentialIndex == credIdx then
+      table.remove(cred_table, index)
+      userIdx = entry.userIndex
+      break
+    end
+  end
+
+  device:emit_event(capabilities.lockCredentials.credentials(cred_table, {visibility = {displayed = false}}))
+  return userIdx
+end
+
+local function delete_credential_from_table_as_user(device, userIdx)
+  -- If User Index is ALL_INDEX, remove all entry from the table
+  if userIdx == ALL_INDEX then
+    device:emit_event(capabilities.lockCredentials.credentials({}, {visibility = {displayed = false}}))
+  end
+
+  -- Get latest credential table
+  local cred_table = device:get_latest_state(
+    "main",
+    capabilities.lockCredentials.ID,
+    capabilities.lockCredentials.credentials.NAME
+  ) or {}
+  local new_cred_table = {}
+
+  -- Re-create credential table
+  for index, entry in pairs(cred_table) do
+    if entry.userIndex ~= userIdx then
+      table.insert(new_cred_table, entry)
+    end
+  end
+
```

Comment:
we can also use deep_copy here!

Reply:
I didn't use deep_copy on purpose here. I think the above way is good because we need to check all entries and exclude credentials with the same user index.

Reply:
I see, good point